sql/core/src/test/scala/org/apache/spark/sql/collation/IndeterminateCollationTestSuite.scala

```diff
@@ -0,0 +1,218 @@
+      sql(s"INSERT INTO $testTableName VALUES ('a', 'b')")
+
+      val expressions = Seq(
+        "c1 = c2",
+        "c1 != c2",
+        "c1 > c2",
+        "STARTSWITH(c1 || c2, c1)",
+        "ENDSWITH(c1 || c2, c2)",
+        "UPPER(c1 || c2) = 'AB'",
+        "INITCAP(c1 || c2) = 'Ab'",
+        "FIND_IN_SET(c1 || c2, 'a,b')",
+        "INSTR(c1 || c2, c1)",
+        "LOCATE(c1, c1 || c2)")
+
+      expressions.foreach { expr =>
+        assertIndeterminateCollationInExpressionError {
+          sql(s"SELECT $expr FROM $testTableName")
+        }
+      }
+    }
+  }
+
+  test("expressions that don't support indeterminate collation and fail in runtime") {
+    withTestTable {
+      sql(s"INSERT INTO $testTableName VALUES ('a', 'b')")
+
+      val expressions = Seq("str_to_map(c1 || c2, 'a', 'b')")
+
+      expressions.foreach { expr =>
+        assertRuntimeIndeterminateCollationError {
+          sql(s"SELECT $expr FROM $testTableName")
+        }
+      }
+    }
+  }
+
+  test("insert works with indeterminate collation") {
+    withTestTable {
+      sql(s"""
+           |INSERT INTO $testTableName
+           |SELECT c1 || c2, c2 || c1
+           |FROM VALUES ('a', 'b') AS t(c1, c2)
+           |""".stripMargin)
+
+      checkAnswer(sql(s"SELECT * FROM $testTableName"), Seq(Row("ab", "ba")))
+    }
+  }
+
+  test("create table as select fails with indeterminate collation") {
+    withTestTable {
+      assertIndeterminateCollationInSchemaError("concat(c1, c2)") {
+        sql(s"""
+             |CREATE TABLE t AS
+             |SELECT c1 || c2 FROM $testTableName
+             |""".stripMargin)
+      }
+
+      assertIndeterminateCollationInSchemaError("col") {
+        sql(s"""
+             |CREATE TABLE t AS
+             |SELECT concat_ws(', ', c1, c2) as col FROM $testTableName
+             |""".stripMargin)
+      }
+
+      assertIndeterminateCollationInSchemaError("arr.element", "map.value", "struct.f1")(sql(s"""
+             |CREATE TABLE t
+             |USING $dataSource
+             |AS SELECT
+             |  array(c1 || c2) AS arr,
+             |  map('a', c1 || c2) AS map,
+             |  named_struct('f1', c1 || c2, 'f2', c2) AS struct
+             |FROM $testTableName
+             |""".stripMargin))
+    }
+  }
+
+  test("can create a view with indeterminate collation") {
+    withTestTable {
+      sql(s"INSERT INTO $testTableName VALUES ('a', 'b')")
+      sql(s"INSERT INTO $testTableName VALUES ('c', 'd')")
+
+      withView("v") {
+        sql(s"""
+             |CREATE VIEW v AS
+             |SELECT c1 || c2 as col FROM $testTableName
+             |""".stripMargin)
+
+        checkAnswer(sql("SELECT * FROM v"), Seq(Row("ab"), Row("cd")))
+        checkAnswer(sql("SELECT DISTINCT COLLATION(col) FROM v"), Seq(Row("null")))
+
+        // group by should fail in runtime when fetching the collator
+        assertRuntimeIndeterminateCollationError {
+          sql(s"SELECT * FROM v GROUP BY col")
+        }
```

Comment:
Should we also, as part of this test case, cover the happy path of fixing the underlying indeterminate collation issue by altering one column of the table to have the same collation as the other column, and then validating that the group by query on top of view succeeds (not sure if the view will need to be altered as well to pick up table changes though)?

Reply:
View will not pick up the new schema if we alter the collation of the column. However we can add test for the happy path by just adding a collate clause in the group by.